confluent_kafka/avro/cached_schema_registry_client.py

```diff
@@ -59,32 +60,30 @@ def __init__(self, url, max_schemas_per_subject=1000):
         # subj => { schema => version }
         self.subject_to_schema_versions = defaultdict(dict)
 
-    def _send_request(self, url, method='GET', body=None, headers=None):
-        if body:
-            body = json.dumps(body)
-            body = body.encode('utf8')
-        _headers = dict()
-        _headers["Accept"] = ACCEPT_HDR
+        self._session = requests.Session()
+        with self._session as s:
+            if ca_location:
+                s.verify = ca_location
+            if (cert_location or key_location) is not None:
+                if (key_location and cert_location) is None:
+                    raise ValueError(
+                        "Both schema.registry.ssl.certificate.location and schema.registry.ssl.key.location"
+                        "must be set: {} {}".format(cert_location, key_location))
+                s.cert = (cert_location, key_location)
+
+    def _send_request(self, url, method='GET', body=None, headers={}):
+        if method not in VALID_METHODS:
+            raise ClientError("Method {} invalid; valid methods include {}".format(method, VALID_METHODS))
+
+        _headers = {'Accept': ACCEPT_HDR}
         if body:
             _headers["Content-Length"] = str(len(body))
             _headers["Content-Type"] = "application/vnd.schemaregistry.v1+json"
+        _headers.update(headers)
 
-        if headers:
-            for header_name in headers:
-                _headers[header_name] = headers[header_name]
-        if method == 'GET':
-            response = requests.get(url, headers=_headers)
-        elif method == 'POST':
-            response = requests.post(url, body, headers=_headers)
-        elif method == 'PUT':
-            response = requests.put(url, body, headers=_headers)
-        elif method == 'DELETE':
-            response = requests.delete(url, headers=_headers)
-        else:
-            raise ClientError("Invalid HTTP request type")
-
-        result = json.loads(response.text)
-        return (result, response.status_code)
+        with self._session as s:
```

Comment:
are you sure the session can be reused after using it in a context like this?

The [docs](http://docs.python-requests.org/en/master/user/advanced/) suggest otherwise:
> This will make sure the session is closed as soon as the with block is exited, even if unhandled exceptions occurred.

Reply:
It closes the connection pool and clears up some state but the session instance is still in play. 

https://github.com/requests/requests/blob/master/requests/sessions.py#L710-L713

The adapter instance itself remains intact although the existing pool manager is cleared as is the adapters current state. 

https://github.com/requests/requests/blob/master/requests/adapters.py#L318-L326

So far as I can tell urllib3 will take care of creating a new pool for us the next time a request is made. 

https://github.com/requests/requests/blob/master/requests/adapters.py#L291-L316

https://github.com/urllib3/urllib3/blob/master/urllib3/poolmanager.py#L266-L279

Either way we aren't leveraging the pooling aspect of sessions which is arguably it's biggest selling point. I'll revert back to using a raw request and persist the CA and Key values in the client instance instead.

Reply:
The session is now left open for the duration of the SR client's existence instead of falling back to raw requests.